js_modules/dagster-ui/packages/ui-core/src/search/AssetSearch.tsx

```diff
@@ -0,0 +1,204 @@
+  | {type: 'complete-secondary'; queryString: string; results: Fuse.FuseResult<SearchResult>[]};
+
+const reducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case 'highlight':
+      return {...state, highlight: action.highlight};
+    case 'change-query':
+      return {...state, queryString: action.queryString, searching: true, highlight: 0};
+    case 'complete-secondary':
+      // If the received results match the current querystring, use them. Otherwise discard.
+      const secondaryResults =
+        action.queryString === state.queryString ? action.results : state.secondaryResults;
+      return {...state, secondaryResults, searching: false};
+    default:
+      return state;
+  }
+};
+
+const initialState: State = {
+  queryString: '',
+  searching: false,
+  secondaryResults: [],
+  highlight: 0,
+  loaded: false,
+};
+
+const DEBOUNCE_MSEC = 100;
+
+type SearchResultGroups = {
+  assetResults: Fuse.FuseResult<SearchResult>[];
+  assetFilterResults: Fuse.FuseResult<SearchResult>[];
+};
+
+function groupSearchResults(secondaryResults: Fuse.FuseResult<SearchResult>[]): SearchResultGroups {
+  return {
+    assetResults: secondaryResults.filter((result) => result.item.type === SearchResultType.Asset),
+    assetFilterResults: secondaryResults.filter((result) =>
+      isAssetFilterSearchResultType(result.item.type),
+    ),
+  };
+}
+
+export const AssetSearch = () => {
+  const history = useHistory();
+  const {loading, searchSecondary} = useGlobalSearch({
+    includeAssetFilters: true,
+  });
+
+  const [state, dispatch] = React.useReducer(reducer, initialState);
+  const {queryString, secondaryResults, highlight} = state;
+
+  const {assetResults, assetFilterResults} = groupSearchResults(secondaryResults);
+
+  const renderedAssetResults = assetResults.slice(0, MAX_ASSET_RESULTS);
+  const renderedFilterResults = assetFilterResults.slice(0, MAX_FILTER_RESULTS);
+
+  const renderedResults = [...renderedAssetResults, ...renderedFilterResults];
+  const numRenderedResults = renderedResults.length;
+
+  const isFirstSearch = React.useRef(true);
+  const firstSearchTrace = React.useRef<null | Trace>(null);
+
+  React.useEffect(() => {
+    if (!loading && secondaryResults) {
+      firstSearchTrace.current?.endTrace();
+    }
+  }, [loading, secondaryResults]);
+
+  const searchAndHandleSecondary = React.useCallback(
+    async (queryString: string) => {
+      const {queryString: queryStringForResults, results} = await searchSecondary(queryString);
+      dispatch({type: 'complete-secondary', queryString: queryStringForResults, results});
+    },
+    [searchSecondary],
+  );
+
+  const debouncedSearch = React.useMemo(() => {
+    const debouncedSearch = debounce(async (queryString: string) => {
+      searchAndHandleSecondary(queryString);
+    }, DEBOUNCE_MSEC);
+    return (queryString: string) => {
+      if (!firstSearchTrace.current && isFirstSearch.current) {
```

Comment:
not sure what this trace logic is, but it existed in the global search dialog so I added similar logic here?